lib/api/CFieldDataCategorizer.cc

```diff
@@ -168,7 +168,9 @@ int CFieldDataCategorizer::computeCategory(const TStrStrUMap& dataRowFields) {
     if (exampleAdded || searchTermsChanged) {
```

Comment:
This condition is the biggest problem with this whole PR.  We only write category definitions when quite a major thing happens to them.  This means the number of matches recorded will be very out-of-date in some cases.

We could go to the opposite extreme and remove this condition altogether.  Then the numbers of matches would always be up-to-date, but it would kill performance as it would mean every single message categorized would lead to a document being indexed in Elasticsearch.  (Most of these documents would overwrite old ones, so the total number of searchable documents would not increase, but the indexing load would be horrific.)

So if we are to record the number of matches for each category then we need a policy on when the numbers will be accurate.  Immediately after a `_flush` or `_close` might be a good one, as that's the policy for when anomalies generated accurately reflect the data posted to a job.  Then we would have implement code to write all categories on `_flush` or `_close`, document that the number of matches is only accurate after a `_flush` or `_close`, and make sure our own UI didn't display numbers of matches that might be highly misleading.

We could possibly also make "number of matches has increased by 10% since the last time this category definition was written" and "this category definition hasn't been written for 3 wall clock hours" extra conditions in this test.  They wouldn't completely solve the consistency problem, but might mean the numbers were good enough for some uses.

The can of worms with number of matches being inconsistent between categories is the reason we haven't included it in the output in the past.  It has been thought about before 😉

Reply:
i agree, this sort of thing came to mind last night as i was falling asleep :). 

I will dig around and experiment. My gut is we periodically update the category if there is a "large" change in `numMatched` and on `_flush` and `_close`.